internal/provider/query_annotation_resource.go

```diff
@@ -0,0 +1,202 @@
+		return
+	}
+
+	var state models.QueryAnnotationResourceModel
+	state.ID = types.StringValue(createdAnnotation.ID)
+	state.Dataset = plan.Dataset
+	state.QueryID = plan.QueryID
+	state.Name = types.StringValue(createdAnnotation.Name)
+	state.Description = types.StringValue(createdAnnotation.Description)
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
+}
+
+func (r *queryAnnotationResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state models.QueryAnnotationResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	dataset := helper.GetDatasetOrAll(state.Dataset)
+
+	var detailedErr client.DetailedError
+	queryAnnotation, err := r.client.QueryAnnotations.Get(ctx, dataset.ValueString(), state.ID.ValueString())
+	if errors.As(err, &detailedErr) {
+		if detailedErr.IsNotFound() {
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.Append(helper.NewDetailedErrorDiagnostic("Reading Honeycomb Query Annotation", &detailedErr))
+		return
+	}
+	if helper.AddDiagnosticOnError(&resp.Diagnostics, "Reading Honeycomb Query Annotation", err) {
+		return
+	}
+
+	state.Name = types.StringValue(queryAnnotation.Name)
+	state.Description = types.StringValue(queryAnnotation.Description)
+	state.QueryID = types.StringValue(queryAnnotation.QueryID)
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
+}
+
+func (r *queryAnnotationResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan models.QueryAnnotationResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	dataset := helper.GetDatasetOrAll(plan.Dataset)
+	queryAnnotation := &client.QueryAnnotation{
+		ID:          plan.ID.ValueString(),
+		Name:        plan.Name.ValueString(),
+		Description: plan.Description.ValueString(),
+		QueryID:     plan.QueryID.ValueString(),
+	}
+
+	updatedAnnotation, err := r.client.QueryAnnotations.Update(ctx, dataset.ValueString(), queryAnnotation)
+	if helper.AddDiagnosticOnError(&resp.Diagnostics, "Updating Honeycomb Query Annotation", err) {
+		return
+	}
+
+	// Update the state with the returned values
+	var state models.QueryAnnotationResourceModel
+	state.ID = plan.ID
+	state.Dataset = plan.Dataset
+	state.QueryID = plan.QueryID
+	state.Name = types.StringValue(updatedAnnotation.Name)
+	state.Description = types.StringValue(updatedAnnotation.Description)
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
+}
+
+func (r *queryAnnotationResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var state models.QueryAnnotationResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	dataset := state.Dataset.ValueString()
```

Comment:
should this use `helper.GetDatasetOrAll` like the rest of the methods here?

Reply:
Oops! I missed that. Fixed.